internal/audits/audits.go

```diff
@@ -0,0 +1,105 @@
+package audits
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/opdev/opcap/internal/operator"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// AuditOption is the function type for passing an option to an audit
+type AuditOption func(options *options) error
+
+// WithSubscription adds a subscription object to the audit
+func WithSubscription(subscription *operator.SubscriptionData) AuditOption {
+	return func(options *options) error {
+		if subscription == nil {
+			return fmt.Errorf("subscription data cannot be nil")
+		}
+		options.Subscription = subscription
+		return nil
+	}
+}
+
+// WithOperatorGroupData adds an operatorgroupdata objec to the audit
+func WithOperatorGroupData(operatorGroupData *operator.OperatorGroupData) AuditOption {
+	return func(options *options) error {
+		if operatorGroupData == nil {
+			return fmt.Errorf("operatorgroupdata cannot be nil")
+		}
+		options.operatorGroupData = operatorGroupData
+		return nil
+	}
+}
+
+// WithNamespace adds a namespace for the audit
+func WithNamespace(namespace string) AuditOption {
+	return func(options *options) error {
+		if namespace == "" {
+			return fmt.Errorf("namespace cannot be empty")
+		}
+		options.namespace = namespace
+		return nil
+	}
+}
+
+// WithClient adds a client to the audit
+func WithClient(client operator.Client) AuditOption {
+	return func(options *options) error {
+		if client == nil {
+			return fmt.Errorf("client cannot be nil")
+		}
+		options.client = client
+		return nil
+	}
+}
+
+// WithTimeout adds a timeout duration to the audit
+func WithTimeout(csvWaitTime int) AuditOption {
+	return func(options *options) error {
+		options.csvWaitTime = time.Duration(csvWaitTime)
+		return nil
+	}
+}
+
+// WithOcpVersion adds the OCP version to the audit
+func WithOcpVersion(ocpVersion string) AuditOption {
+	return func(options *options) error {
+		options.OcpVersion = ocpVersion
+		return nil
+	}
+}
+
+type options struct {
+	Subscription      *operator.SubscriptionData
+	operatorGroupData *operator.OperatorGroupData
+	namespace         string
+	client            operator.Client
+	CsvTimeout        bool
+	csvWaitTime       time.Duration
+	Csv               v1alpha1.ClusterServiceVersion
+	OcpVersion        string
+	customResources   []map[string]interface{}
+	operands          []unstructured.Unstructured
+}
+
+type auditFn func(context.Context) error
+
+// New returns a function corresponding to a passed in audit plan
+func New(ctx context.Context, auditType string, opts ...AuditOption) auditFn {
+	switch strings.ToLower(auditType) {
+	case "operatorinstall":
+		return operatorInstall(ctx, opts...)
+	case "operatorcleanup":
+		return operatorCleanUp(ctx, opts...)
+	case "operandinstall":
+		return operandInstall(ctx, opts...)
+	case "operandcleanup":
+		return operandCleanUp(ctx, opts...)
+	}
+	return nil
+}
```

Comment:
This is pretty neat. Although I really think the auditPlan thing must be revamped. I don't think cleanup functions should be part of that input. If you install an operator the function should know how to cleanup without involving the user. Not a blocker though. But if by any way you come up with a solution for this will be very welcome.

Reply:
> This is pretty neat. Although I really think the auditPlan thing must be revamped. I don't think cleanup functions should be part of that input. If you install an operator the function should know how to cleanup without involving the user. Not a blocker though. But if by any way you come up with a solution for this will be very welcome.

I agree. I did it this way to maintain backward compat, but I'm thinking that needs to be different. I will give it a thought and see what could be done.

Reply:
I think in another PR though.